Update Oracle connector documentation (#5014)
prestosql/presto: presto-docs/src/main/sphinx/connector/oracle.rst

Oracle Connector
================

The Oracle connector allows querying and creating tables in an external Oracle

> database. Connectors let Presto join data provided by different databases,

Author: Most of the intros for the connectors start with a similar opening, it's mostly to orient people to where they are in the docs.

Oracle Connector
================

The Oracle connector allows querying and creating tables in an external Oracle
database. It can join data between different systems, like Oracle and Hive, or
between different Oracle database instances.

Configuration
-------------

To configure the Oracle connector, create a catalog properties file in
``etc/catalog`` named ``oracle.properties``. 

To mount the Oracle connector as the ``oracle`` catalog, create a file with the

> following connection properties:

Author: I drafted an edited version of your suggestion.

        infinity.
      - ``HALF_DOWN`` - Rounding mode to round towards
        ``nearest neighbor`` unless both neighbors are
        equidistant, in which case rounding down is used.
      - ``HALF_EVEN`` - Rounding mode to round towards the
        ``nearest neighbor`` unless both neighbors are equidistant,
        in which case rounding towards the even neighbor is
        performed.
      - ``HALF_UP`` - Rounding mode to round towards
        ``nearest neighbor`` unless both neighbors are
        equidistant, in which case rounding up is used
      - ``UP`` - Rounding mode to round towards zero.
      - ``DOWN`` - Rounding mode to round towards zero.

    - ``UNNECESSARY``

Synonyms
--------

Based on performance reasons, Presto disables support for Oracle ``SYNONYM``. To
include ``SYNONYM``, add the following configuration property: 

.. code-block:: none

    oracle.synonyms.enabled=true

Pushdown
--------

The connector supports :doc:`pushdown </optimizer/pushdown>` for optimized query processing.

Limitations
-----------

The following SQL statements are not supported:

* :doc:`/sql/delete`
* :doc:`/sql/alter-table`

> * :doc:`/sql/create-table` (:doc:`/sql/create-table-as` is supported)

Author: @losipiuk can you confirm that `Create table` isn't supported and that `create tables as` is supported? David asked me to double check. Thanks!

Reviewer: Just checked. Create table is also supported.
@findepi do we have a test coverage for that? Quickly I could not find any.

Something as simple as
```java
    @Test
    public void testCreateTable()
    {
        getQueryRunner().execute("CREATE TABLE t (c BIGINT)");
        getQueryRunner().execute("INSERT INTO t VALUES(10)");
        assertThat(query("SELECT CAST(c as VARCHAR) from t")).matches("VALUES (VARCHAR '10')");
    }
```

or 

```java
    @Test
    public void testCreateTable()
    {
        getQueryRunner().execute("CREATE TABLE t (c BIGINT)");
        assertThat(query("DESCRIBE t")).matches(results -> results.getRowCount() > 0);
    }
```

Reviewer 2: `io.prestosql.testing.AbstractTestDistributedQueries#testCreateTable`

Author: @losipiuk thanks so much! I'm going to create a PR to remove `create tables` and the note about `create tables as` from the limitations section.

Reviewer: > io.prestosql.testing.AbstractTestDistributedQueries#testCreateTable

Cool. I was looking in the `*Smoke*` ones.

   TABLE <table> AS SELECT`` operations. When data is inserted into existing
   tables ``Oracle to Presto`` type mapping is used.

.. list-table:: Presto to Oracle Type Mapping
  :widths: 20, 20, 60
  :header-rows: 1

  * - Presto type
    - Oracle database type
    - Notes
  * - ``TINYINT``
    - ``NUMBER(3)``
    -
  * - ``SMALLINT``
    - ``NUMBER(5)``
    -
  * - ``INTEGER``
    - ``NUMBER(10)``
    -
  * - ``BIGINT``
    - ``NUMBER(19)``
    -
  * - ``DECIMAL(p, s)``
    - ``NUMBER(p, s)``
    -
  * - ``REAL``
    - ``BINARY_FLOAT``
    -
  * - ``DOUBLE``
    - ``BINARY_DOUBLE``
    -
  * - ``VARCHAR``
    - ``NCLOB``
    -
  * - ``VARCHAR(n)``
    - ``VARCHAR2(n CHAR)`` or ``NCLOB``
    - See :ref:`character mapping`
  * - ``CHAR(n)``
    - ``CHAR(n CHAR)`` or ``NCLOB``
    - See :ref:`character mapping`
  * - ``VARBINARY``
    - ``BLOB``
    -
  * - ``DATE``
    - ``DATE``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP``
    - ``TIMESTAMP(3)``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP WITH TIME ZONE``
    - ``TIMESTAMP(3) WITH TIME ZONE``
    - See :ref:`datetime mapping`

.. _number mapping:

Mapping numeric types
^^^^^^^^^^^^^^^^^^^^^

> An Oracle ``NUMBER(p, s)`` maps to Presto's ``DECIMAL(p, s)`` except in these

Reviewer: This section renders incorrectly:

![image](https://user-images.githubusercontent.com/890111/92028881-0899d300-ed65-11ea-85f0-740616efc284.png)

Author: Fixed

    connection-user=root
    connection-password=secret

By default, the Oracle connector uses connection pooling for performance
improvement. The below configuration shows the typical default values. To update
them, change the properties in the catalog configuration file:

.. code-block:: properties

    oracle.connection-pool.max-size=30
    oracle.connection-pool.min-size=1
    oracle.connection-pool.inactive-timeout=20m

> To disable connection pooling, update properties to include:

Reviewer 3: "to include the following"

.. code-block:: none

    oracle.connection-pool.enabled=false

.. note::

>     Include catalog files on all of your nodes, this requires you to restart.

Reviewer 3: This is two sentences separated with a comma

Multiple Oracle servers
^^^^^^^^^^^^^^^^^^^^^^^

If you want to connect to multiple Oracle servers, configure another instance of
the Oracle connector as a separate catalog.

To add another Oracle catalog, create a new properties file. For example, if
you name the property file ``sales.properties``, Presto creates a catalog named
sales.

Querying Oracle
---------------

> The Oracle connector provides a schema for every Oracle database. Use the below

Reviewer 3: This last sentence is misleading, since it's not clear what "below queries" refers to. `SHOW SCHEMAS` will always work. The rest are examples, not something you could directly use.

Author: I restructured this section, hopefully its more clear.

queries to test your setup.

If you used a different name for your catalog properties file, use that catalog
name instead of ``oracle``.

.. note::
    The user configuration, in the connection properties file, determines your

>     privileges in these schemas. If the Oracle user doesn't have access to a

Reviewer 3: The sentence seems clearer if we reverse it

> The Oracle user must have access to the table in order to access it from Presto.

    table, you can't access it through Presto either.

Run ``SHOW SCHEMAS`` to see the available Oracle databases::

    SHOW SCHEMAS FROM oracle;

If you have an Oracle database named ``web``, run ``SHOW TABLES`` to see the
tables it contains::

    SHOW TABLES FROM oracle.web;

To see a list of the columns in the ``clicks`` table in the ``web``

> database run either of the following::

Reviewer 3: Missing comma

    DESCRIBE oracle.web.clicks;
    SHOW COLUMNS FROM oracle.web.clicks;

> To access the ``clicks`` table in the ``web`` database run::

Reviewer 3: The original structure seems better. It looks strange to have "run" at the end of the sentence, since it's missing a comma and seems to be missing something. The correct way to write a sentence like this would be

> To access the clicks table in the web database, run the following:

    SELECT * FROM oracle.web.clicks;

Mapping data types between Presto and Oracle
--------------------------------------------

> Oracle supports types that Presto doesn't and Presto supports types that Oracle

Reviewer 3: How about

> Both Oracle and Presto have types that are not supported by the Oracle connector.

Presto data type mapping:

.. list-table:: Oracle to Presto type mapping
  :widths: 20, 20, 60
  :header-rows: 1

  * - Oracle database type
    - Presto type
    - Notes
  * - ``NUMBER(p, s)``
    - ``DECIMAL(p, s)``
    -  See :ref:`number mapping`
  * - ``NUMBER(p)``
    - ``DECIMAL(p, 0)``
    - See :ref:`number mapping`
  * - ``FLOAT[(p)]``
    - ``DOUBLE``
    -
  * - ``BINARY_FLOAT``
    - ``REAL``
    -
  * - ``BINARY_DOUBLE``
    - ``DOUBLE``
    -
  * - ``VARCHAR2(n CHAR)``
    - ``VARCHAR(n)``
    -
  * - ``VARCHAR2(n BYTE)``
    - ``VARCHAR(n)``
    -
  * - ``NVARCHAR2(n)``
    - ``VARCHAR(n)``
    -
  * - ``CHAR(n)``
    - ``CHAR(n)``
    -
  * - ``NCHAR(n)``
    - ``CHAR(n)``
    -
  * - ``CLOB``
    - ``VARCHAR``
    -
  * - ``NCLOB``
    - ``VARCHAR``
    -
  * - ``RAW(n)``
    - ``VARBINARY``
    -
  * - ``BLOB``
    - ``VARBINARY``
    -
  * - ``DATE``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p)``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p) WITH TIME ZONE``
    - ``TIMESTAMP WITH TIME ZONE``
    - See :ref:`datetime mapping`

> Use the default configureation property, ``unsupported-type.handling``, to specify

Reviewer: ```suggestion
Use the ``unsupported-type.handling`` configuration property, to specify
```

Author: added the edit to my bulk edits.

      Oracle ``NUMBER`` data type specifies higher scale than is supported in 
      Presto. Possible values are:

      - ``UNNECESSARY`` - Rounding mode to assert that the 
        requested operation has an exact result, 
        hence no rounding is necessary.
      - ``CEILING`` - Rounding mode to round towards
        positive infinity.
      - ``FLOOR`` - Rounding mode to round towards negative
        infinity.
      - ``HALF_DOWN`` - Rounding mode to round towards
        ``nearest neighbor`` unless both neighbors are
        equidistant, in which case rounding down is used.
      - ``HALF_EVEN`` - Rounding mode to round towards the
        ``nearest neighbor`` unless both neighbors are equidistant,
        in which case rounding towards the even neighbor is
        performed.
      - ``HALF_UP`` - Rounding mode to round towards
        ``nearest neighbor`` unless both neighbors are
        equidistant, in which case rounding up is used
      - ``UP`` - Rounding mode to round towards zero.
      - ``DOWN`` - Rounding mode to round towards zero.

    - ``UNNECESSARY``

Synonyms
--------

Based on performance reasons, Presto disables support for Oracle ``SYNONYM``. To

> include ``SYNONYM``, add the following: 

Reviewer 3: "add the following configuration property"

.. code-block:: none

    oracle.synonyms.enabled=true

Pushdown
--------

The connector supports :doc:`pushdown </optimizer/pushdown>` for optimized query

> processing.

Reviewer 3: Don't wrap a single word as it looks funny and makes it harder to read. There is no strict line length limit.

Limitations
-----------

> The following SQL statements aren't supported:

Reviewer 3: Using a contraction here looks weird, since this is a formal statement. Contractions are fine in prose, but for a short single statement like this, I don't see a benefit to using it.

Author: I was using contractions based on the Google Style Guide: https://developers.google.com/style/contractions

> Negation contractions
In particular, it's fine to use -n't contractions, such as isn't, don't, and can't.

One reason that such contractions are useful is that it's sometimes easy for a reader to miss the word **not**, whereas it's harder to misread don't as do.

    -
  * - ``NCLOB``
    - ``VARCHAR``
    -
  * - ``RAW(n)``
    - ``VARBINARY``
    -
  * - ``BLOB``
    - ``VARBINARY``
    -
  * - ``DATE``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p)``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p) WITH TIME ZONE``
    - ``TIMESTAMP WITH TIME ZONE``
    - See :ref:`datetime mapping`

If an Oracle table uses a type not listed in the above table, then you can use the
``unsupported-type.handling`` configuration property to specify Presto behavior.
For example:

- If ``unsupported-type.handling`` is set to ``FAIL``, then the
  querying of an unsupported table fails.
- If ``unsupported-type.handling`` is set to ``IGNORE``, 
  then you can't see the unsupported types in Presto.
- If ``unsupported-type.handling`` is set to ``CONVERT_TO_VARCHAR``, 
  then the column is exposed as unbounded ``VARCHAR``.

Presto to Oracle type mapping
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Presto supports creating tables with the following types in an Oracle database.
The table shows the mappings from Presto to Oracle data types:

.. note::
   For types not listed in the table below, Presto can't perform the ``CREATE

>    TABLE <table> AS SELECT`` operations. When data is inserted into existing

Reviewer 4: This sentence was suggested by @losipiuk .. however I am not sure this actually true. At least all other connector docs suggest that the Presto to Oracle mapping is used for ALL write operations .. and not just for CTAS ... 

Can you clarify if the Oracle connector is just different or if all other docs are wrong or whatever is technically correct @losipiuk @findepi and @electrum

Reviewer: The "Presto to <some DB>" mapping is used only in CTAS flow, when the external table does not yet exist, and we need to build the a schema for it based on Presto types of data used in CTAS statement.
For query operations as well as INSERT operations the "<some DB> to Presto" mapping is used. We already have columns in external database and we map types of those columns to Presto types.
This is not Oracle specific, but true for all connectors based on presto-base-jdbc.

Still I am not at all defending the wording I suggested. It is highly probable that one can phrase it much better :)

Hera are the code pointers:
* CTAS: https://github.com/prestosql/presto/blob/fd66bb72ddb9cfc311511900b1454cef385f172d/presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcPageSink.java#L74
* INSERT: https://github.com/prestosql/presto/blob/fd66bb72ddb9cfc311511900b1454cef385f172d/presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcPageSink.java#L89

Reviewer 4: So looking at the code I conclude that this is true for all connectors that use the base-jdbc stuff .. so including mysql, postgresql, oracle, sqlserver and maybe others. 

At this stage I suggest we leaven the wording as it is in this PR and create another PR that clarifies this and corrects the docs for all affected connectors. Do you agree with this approach @losipiuk  and @electrum  .. if yes .. can we get this PR merged.

Reviewer: SGTM

    connection-user=root
    connection-password=secret

The below configuration shows the typical default values. To update them, change
the properties in the catalog configuration file:

.. code-block:: properties

    oracle.connection-pool.max-size=30
    oracle.connection-pool.min-size=1
    oracle.connection-pool.inactive-timeout=20m

By default, the Oracle connector uses connection pooling for performance
improvement. To disable connection pooling, update properties to include:

.. code-block:: none

>     oracle.connection-pool.enabled=false

Author: Added to edits.

    - ``CHAR(n)``
    -
  * - ``NCHAR(n)``
    - ``CHAR(n)``
    -
  * - ``CLOB``
    - ``VARCHAR``
    -
  * - ``NCLOB``
    - ``VARCHAR``
    -
  * - ``RAW(n)``
    - ``VARBINARY``
    -
  * - ``BLOB``
    - ``VARBINARY``
    -
  * - ``DATE``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p)``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p) WITH TIME ZONE``
    - ``TIMESTAMP WITH TIME ZONE``
    - See :ref:`datetime mapping`

If an Oracle table contains a type that's not listed in the above table, an
error shows:

 - An ``unsupported-type.handling-strategy`` equals ``FAIL``
   (default configuration), any queries on the table fail.
 - An ``unsupported-type.handling-strategy`` equals ``IGNORE``
   (default configuration), Presto no longer accesses the column.
 - An ``unsupported-type.handling-strategy`` equals ``CONVERT_TO_VARCHAR``
   (default configuration), a column value converts to an unbounded ``VARCHAR``.

Presto to Oracle type mapping
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Presto supports creating tables with the following types in an Oracle database.
The table shows the mappings from Presto to Oracle data types:

.. note::

>    For types not listed in the table below, Presto can't perform the ``CREATE
>    TABLE`` and ``AS SELECT``operations to insert info into existing
>    tables that use ``Oracle to Presto`` type mapping.

Reviewer: this is single operation `CREATE TABLE <some_table> AS SELECT ....`

```suggestion
   For types not listed in the table below, Presto can't perform the ``CREATE
   TABLE <table> AS SELECT`` operations. 
   When data is inserted into existing tables ``Oracle to Presto`` type mapping is used.
```

Author: Added to edits.

    - ``BLOB``
    -
  * - ``DATE``
    - ``DATE``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP``
    - ``TIMESTAMP(3)``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP WITH TIME ZONE``
    - ``TIMESTAMP(3) WITH TIME ZONE``
    - See :ref:`datetime mapping`

.. _number mapping:

Mapping numeric types
^^^^^^^^^^^^^^^^^^^^^

An Oracle ``NUMBER(p, s)`` maps to Presto's ``DECIMAL(p, s)`` except in these
conditions:

 - No precision is specified for the column (example: ``NUMBER`` or
   ``NUMBER(*)``), unless ``oracle.number.default-scale`` is set. 
 - Scale (``s`` ) is greater than precision. 
 - Precision (``p`` ) is greater than 38. 
 - Scale is negative and the difference between ``p`` and ``s`` is greater than
   38, unless ``oracle.number.rounding-mode`` is set to a different value than
   ``UNNECESSARY``. 

If a column meets any of the conditions:

 - An ``unsupported-type.handling-strategy`` equals ``FAIL``
   (default configuration) any queries on the table fail.
 - An ``unsupported-type.handling-strategy`` equals ``IGNORE``
   (default configuration), Presto no longer accesses the column.
 - An ``unsupported-type.handling-strategy`` equals ``CONVERT_TO_VARCHAR``

>    (default configuration), a column value converts to an unbounded ``VARCHAR``.

Reviewer: Please remove this one.
It is not related for `Presto to Oracle Type Mapping`. I think someone copy pasted this initially incidentally.

Author: Removed.

      - ``IGNORE`` - column is not accessible.
      - ``CONVERT_TO_VARCHAR`` - column is converted to unbounded ``VARCHAR``.

    - ``IGNORE``
  * - ``oracle.number.default-scale``
    - ``number_default_scale``               
    - Default Presto ``DECIMAL`` scale for Oracle ``NUMBER`` (without precision 
      and scale) date type. When not set then such column is treated as not 
      supported.
    - not set
  * - ``oracle.number.rounding-mode``
    - ``number_rounding_mode``
    - Rounding mode for the Oracle ``NUMBER`` data type. This is useful when 
      Oracle ``NUMBER`` data type specifies higher scale than is supported in 
      Presto. Possible values are:

      - ``UNNECESSARY`` - Rounding mode to assert that the 
        requested operation has an exact result, 
        hence no rounding is necessary.
      - ``CEILING`` - Rounding mode to round towards
        positive infinity.
      - ``FLOOR`` - Rounding mode to round towards negative
        infinity.
      - ``HALF_DOWN`` - Rounding mode to round towards
        ``nearest neighbor`` unless both neighbors are
        equidistant, in which case rounding down is used.

>       - ``HALF_EVEN`` - Rounding mode to round towards the
>         ``nearest neighbor`` unless both neighbors are
>         in which case rounding towards the nearest neighbor is
>         performed.

Reviewer: I made suggestion fix here before. The rounding here is done differently.
```suggestion
      - ``HALF_EVEN`` - Rounding mode to round towards the
        ``nearest neighbor`` unless both neighbors are
        equidistant, in which case rounding towards the even neighbor is
        performed.
```

Author: ooh, I actually did fix that but looking back at the doc, there is more than one instance where similar language is used.

    connection-user=root
    connection-password=secret

Include catalog files on all of your nodes, this requires you to restart.

By default, the Oracle connector uses connection pooling for performance
improvement.

The below configuration shows the typical default values. To update them, change
the properties in the catalog configuration file:

.. code-block:: properties

    oracle.connection-pool.max-size=30
    oracle.connection-pool.min-size=1
    oracle.connection-pool.inactive-timeout=20m

To disable connection pooling, update properties to include:

.. code-block:: none

    oracle.connection-pool.enabled=false

Multiple Oracle servers
^^^^^^^^^^^^^^^^^^^^^^^

If you want to connect to multiple Oracle servers, configure another instance of
the Oracle plugin as a separate catalog.

To add another Oracle catalog, create a new properties file. For example, if
you name the property file ``sales.properties``, Presto creates a catalog named
sales.

Querying Oracle
---------------

> The Oracle connector provides a schema for every Oracle server. Use the below

Author: Thanks for the explanation, it was helpful.

queries to test your setup.

If you used a different name for your catalog properties file, use that catalog
name instead of ``oracle``.

.. note::
    The user configuration, in the connection properties file, determines your
    privileges in these schemas. If the user doesn't have access to these
    tables, you can't access them either.

Run ``SHOW SCHEMAS`` to see the available Oracle databases::

    SHOW SCHEMAS FROM oracle;

If you have an Oracle database named ``web``, run ``SHOW TABLES`` to see the
tables it owns::

    SHOW TABLES FROM oracle.web;

To see a list of the columns in the ``clicks`` table in the ``web``
database run either of the following::

    DESCRIBE oracle.web.clicks;
    SHOW COLUMNS FROM oracle.web.clicks;

To access the ``clicks`` table in the ``web`` database run::

    SELECT * FROM oracle.web.clicks;

Mapping data types between Presto and Oracle
--------------------------------------------

Oracle and Presto support types that the other doesn't, the following tables
show their type mapping:

> .. list-table:: Type mapping properties

Reviewer: Maybe create a specific section for this table:
```
Type mapping configuration properties
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Also it feels to me that it would make sense to make it the last section.
After `Oracle to Presto type mapping` and `Presto to Oracle type mapping`

Presto data type mapping:

.. list-table:: Oracle to Presto type mapping
  :widths: 20, 20, 60
  :header-rows: 1

  * - Oracle database type
    - Presto type
    - Notes
  * - ``NUMBER(p, s)``
    - ``DECIMAL(p, s)``
    -  See :ref:`number mapping`
  * - ``NUMBER(p)``
    - ``DECIMAL(p, 0)``
    - See :ref:`number mapping`
  * - ``FLOAT[(p)]``
    - ``DOUBLE``
    -
  * - ``BINARY_FLOAT``
    - ``REAL``
    -
  * - ``BINARY_DOUBLE``
    - ``DOUBLE``
    -
  * - ``VARCHAR2(n CHAR)``
    - ``VARCHAR(n)``
    -
  * - ``VARCHAR2(n BYTE)``
    - ``VARCHAR(n)``
    -
  * - ``NVARCHAR2(n)``
    - ``VARCHAR(n)``
    -
  * - ``CHAR(n)``
    - ``CHAR(n)``
    -
  * - ``NCHAR(n)``
    - ``CHAR(n)``
    -
  * - ``CLOB``
    - ``VARCHAR``
    -
  * - ``NCLOB``
    - ``VARCHAR``
    -
  * - ``RAW(n)``
    - ``VARBINARY``
    -
  * - ``BLOB``
    - ``VARBINARY``
    -
  * - ``DATE``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p)``
    - ``TIMESTAMP``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP(p) WITH TIME ZONE``
    - ``TIMESTAMP WITH TIME ZONE``
    - See :ref:`datetime mapping`

> If an Oracle table contains a type that's not listed in the above table, an

Reviewer: This one is rendered weirdly with extra indentatinon. Not sure how to fix that.

![image](https://user-images.githubusercontent.com/890111/91697416-61a11580-eb71-11ea-967e-3ec110aab7a9.png)

Author: I'm not sure how to fix this but I'm working on it. It's structured the same way the other lists in the document are structured.

Reviewer: The problem that it is hard to follow (at least to me) that it says about different possible outcomes, depending on value of configuration property `unsupported-type.handling-strategy`.
Also I have just noticed that each option is marked with "(default configuration)" :).
Also it should be `unsupported-type.handling` (without `-strategy`); The version with `-strategy` would work too, but it is legacy and will be removed eventually.

I would expect something like:

```
You can specify Presto behaviour for situation when type not listed above is used in Oracle table by setting `unsupported-type.handling-strategy` configuration property.

If ``unsupported-type.handling`` is set to ``FAIL``  querying table with unsupported table will fail.

If ``unsupported-type.handling`` is set to ``IGNORE`` (default configuration) columns with unsupported types will not be exposed in Presto.

If ``unsupported-type.handling`` is set to ``CONVERT_TO_VARCHAR`` column will be exposed as unbounded VARCHAR
```

Of course please make it more English :)

> If an Oracle table contains a type that's not listed in the above table, an

Reviewer: This section requires major rephrasing

error shows:

 - An ``unsupported-type.handling-strategy`` equals ``FAIL``
   (default configuration), any queries on the table fail.
 - An ``unsupported-type.handling-strategy`` equals ``IGNORE``
   (default configuration), Presto no longer accesses the column.
 - An ``unsupported-type.handling-strategy`` equals ``CONVERT_TO_VARCHAR``
   (default configuration), a column value converts to an unbounded ``VARCHAR``.

Presto to Oracle type mapping
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Presto supports creating tables with the following types in an Oracle database.
The table shows the mappings from Presto to Oracle data types:

.. note::

>    Oracle connectors can't write Presto types that aren't listed in the table.

Reviewer: I think it should say that for types not listed below Presto cannot perform `CREATE TABLE ... AS SELECT ..` operation. Inserting into alerady existing table uses `Oracle to Presto` type mapping.

cc: @findepi are there any quirks here?

Author: I edited based on your note, lemme know what you think of the new version.

    - ``NUMBER(p, s)``
    -
  * - ``REAL``
    - ``BINARY_FLOAT``
    -
  * - ``DOUBLE``
    - ``BINARY_DOUBLE``
    -
  * - ``VARCHAR``
    - ``NCLOB``
    -
  * - ``VARCHAR(n)``
    - ``VARCHAR2(n CHAR)`` or ``NCLOB``
    - See :ref:`character mapping`
  * - ``CHAR(n)``
    - ``CHAR(n CHAR)`` or ``NCLOB``
    - See :ref:`character mapping`
  * - ``VARBINARY``
    - ``BLOB``
    -
  * - ``DATE``
    - ``DATE``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP``
    - ``TIMESTAMP(3)``
    - See :ref:`datetime mapping`
  * - ``TIMESTAMP WITH TIME ZONE``
    - ``TIMESTAMP(3) WITH TIME ZONE``
    - See :ref:`datetime mapping`

.. _number mapping:

Mapping numeric types
^^^^^^^^^^^^^^^^^^^^^

An Oracle ``NUMBER(p, s)`` maps to Presto's ``DECIMAL(p, s)`` except in these
conditions:

 - No precision is specified for the column (example: ``NUMBER`` or
   ``NUMBER(*)``), unless ``oracle.number.default-scale`` is set. 
 - Scale (``s`` ) is greater than precision. 
 - Precision (``p`` ) is greater than 38. 
 - Scale is negative and the difference between ``p`` and ``s`` is greater than
   38, unless ``oracle.number.rounding-mode`` is set to a different value than
   ``UNNECESSARY``. 

> If a column meets any of these conditions, it shows an error:

Reviewer: again. I have hard time understanding this one.

 - An ``unsupported-type.handling-strategy`` equals ``FAIL``
   (default configuration) any queries on the table fail.
 - An ``unsupported-type.handling-strategy`` equals ``IGNORE``
   (default configuration), Presto no longer accesses the column.
 - An ``unsupported-type.handling-strategy`` equals ``CONVERT_TO_VARCHAR``
   (default configuration), a column value converts to an unbounded ``VARCHAR``.

If ``s`` is negative, ``NUMBER(p, s)`` maps to ``DECIMAL(p + s, 0)``.

For Oracle ``NUMBER`` (without precision and scale), you can change
``oracle.number.default-scale=s`` and map the column to ``DECIMAL(38, s)``.

.. _datetime mapping:

Mapping datetime types
^^^^^^^^^^^^^^^^^^^^^^

Selecting a timestamp with fractional second precision (``p``) greater than 3

> truncates the fractional seconds to three digits instead of rounding it.

Reviewer: I would drop `instead of rounding it`.

> Oracle ``DATE`` values may store hours, minutes, and seconds, mapped
> to Presto ``TIMESTAMP``.

Reviewer: ```suggestion
Oracle ``DATE`` type may store hours, minutes, and seconds, so it is mapped
to Presto ``TIMESTAMP``.
```

4000. A larger or unbounded ``VARCHAR`` maps to ``NCLOB``.

Presto's ``CHAR(n)`` maps to ``CHAR(n CHAR)`` if ``n`` is no greater than 2000.
A larger ``CHAR`` maps to ``NCLOB``.

Using ``CREATE TABLE AS`` to create an ``NCLOB`` column from a ``CHAR`` value
removes the trailing spaces from the initial values for the column. Inserting
``CHAR`` values into existing ``NCLOB`` columns keeps the trailing spaces. For
example::

    presto> CREATE TABLE vals AS SELECT CAST('A' as CHAR(2001)) col;
    presto> INSERT INTO vals (col) VALUES (CAST('BB' as CHAR(2001)));
    presto> SELECT LENGTH(col) FROM vals;
    2001
    1

Attempting to write a ``CHAR`` that doesn't fit in the column's actual size
fails. This is also true for the equivalent ``VARCHAR`` types.

.. note:: 

    This behaviors above may change in a future version.

Synonyms
--------

Based on performance reasons, Presto disables support for Oracle ``SYNONYM``. To
include ``SYNONYM``, add the following: 

.. code-block:: none

    oracle.synonyms.enabled=true

Pushdown
--------

The connector supports :doc:`pushdown </optimizer/pushdown>` for optimized query
processing.

Limitations
-----------

> Presto supports connecting to Oracle Database 12c. The following SQL statements

Reviewer 5: 12c is not the only version, 11g works, maybe others too

Author: Removed reference in new draft.
